### DISCUSSION TITLE
How to get request post data OnBeforePopup

### DISCUSSION THREAD

**Discussion by GITMoor** [cefsharp/CefSharp#4277] [Q&A] [answered]
When creating a new tab, sometimes its no enough to just pass the url to the new browser instance. So, is there a way to retrieve post data, headers from request to pass then into a new browser instance?

**Comment by amaitland** [comment, marked as answer]
> So, is there a way to retrieve post data, headers from request to pass then into a new browser instance?

No. If you are canceling popup creation then that's not possible. 

If you are using WinForms then you can use code like the example shown in https://github.com/cefsharp/CefSharp/pull/3529#issuecomment-828934239

**Comment by GITMoor** [comment]
Yeah, that looks promising, but i noticed that if you close tab with hosting ChromiumWebBrowser all popups in other tabs will continue to popup in new window, is there a way to make them pop up as before closing the main tab or i need to prevent closing main tab?

**Reply by amaitland** [reply]
When you dispose the `ChromiumWebBrowser` then the LifeSpanHandler reference is freed. The LifeSpanHandler opens the popups in the `ChromiumHostControl`, so without it they open by default in a new window.

You don't need to keep the tab open, you'd just need to stop the `ChromiumWebBrowser` from being disposed. So remove it from it's parent and `Dispose` of it later when all popups are closed.

I've not looked into this scenario to see if it's possible to provide better support than that. You'd need to add a breakpoint to https://github.com/cefsharp/CefSharp/blob/cefsharp/106/CefSharp.Core.Runtime/Internals/ClientAdapter.cpp#L117 and see if it was still called after the `ChromiumWebBrowser` had been disposed to know if any improvements could be made.

**Reply by GITMoor** [reply]
Setting ChromiumWebBrowser.Parent to null before disposing its parent form works for me, thanks!
